drivers/i3c/i3c_stm32.c

```diff
@@ -0,0 +1,2129 @@
+	if (ret != 0) {
+		LOG_ERR("Addr slots init fail, err=%d", ret);
+		return ret;
+	}
+
+	config->irq_config_func(dev);
+	i3c_stm32_configure(dev, I3C_CONFIG_CONTROLLER, NULL);
+
+#ifdef CONFIG_PM_DEVICE_RUNTIME
+	(void)pm_device_runtime_enable(dev);
+#endif
+
+	/* Perform bus initialization only if there are devices that already exist on the bus */
+	if (config->drv_cfg.dev_list.num_i3c > 0) {
+		ret = i3c_bus_init(dev, &config->drv_cfg.dev_list);
+		if (ret != 0) {
+			LOG_ERR("Failed to do i3c bus init, err=%d", ret);
+			return ret;
+		}
+	}
+
+	return 0;
+}
+
+static void i3c_stm32_event_isr_tx(const struct device *dev)
+{
+	const struct i3c_stm32_config *config = dev->config;
+	struct i3c_stm32_data *data = dev->data;
+	I3C_TypeDef *i3c = config->i3c;
+
+	switch (data->msg_state) {
+	case STM32_I3C_MSG: {
+		uint8_t *buf = NULL;
+		size_t *offset = NULL;
+		uint32_t len = 0;
+
+		i3c_stm32_curr_msg_xfer_get_buf(dev, &buf, &len, &offset);
+
+		if (i3c_stm32_fill_tx_fifo(dev, buf, len, offset)) {
+			i3c_stm32_curr_msg_xfer_next(dev);
+		}
+
+		break;
+	}
+	case STM32_I3C_MSG_DAA: {
+		struct i3c_device_desc *target;
+		uint8_t bcr;
+		uint8_t dcr;
+		uint8_t dyn_addr = 0;
+		int ret;
+
+		bcr = (data->pid >> 8) & 0xFF;
+		dcr = data->pid & 0xFF;
+		data->pid >>= 16;
+
+		/* Find the device in the device list */
+		ret = i3c_dev_list_daa_addr_helper(&data->drv_data.attached_dev.addr_slots,
+						   &config->drv_cfg.dev_list, data->pid, false,
+						   false, &target, &dyn_addr);
+		if (ret != 0) {
+			/* TODO: figure out what is the correct sequence to exit form this error
+			 * It is expected that a TX overrun error to occur which triggers err isr
+			 */
+			LOG_ERR("No dynamic address could be assigned to target");
+
+			return;
+		}
+
+		/* Put the new dynamic address in TX FIFO for transmission */
+		LL_I3C_TransmitData8(i3c, dyn_addr);
+
+		if (target != NULL) {
+			/* Update target descriptor */
+			target->dynamic_addr = dyn_addr;
+			target->bcr = bcr;
+			target->dcr = dcr;
+		}
+
+		/* Mark the address as used */
+		i3c_addr_slots_mark_i3c(&data->drv_data.attached_dev.addr_slots, dyn_addr);
+
+		/* Mark the static address as free */
+		if ((target != NULL) && (target->static_addr != 0) &&
+		    (dyn_addr != target->static_addr)) {
+			i3c_addr_slots_mark_free(&data->drv_data.attached_dev.addr_slots, dyn_addr);
+		}
+
+		break;
+	}
+	case STM32_I3C_MSG_CCC: {
+		struct i3c_ccc_payload *payload = data->ccc_payload;
+
+		if (payload->ccc.num_xfer < payload->ccc.data_len) {
+			LL_I3C_TransmitData8(i3c, payload->ccc.data[payload->ccc.num_xfer++]);
+		}
+		break;
+	}
+	case STM32_I3C_MSG_CCC_P2: {
+		struct i3c_ccc_target_payload *target = data->ccc_target_payload;
+
+		if (target->num_xfer < target->data_len) {
+			LL_I3C_TransmitData8(i3c, target->data[target->num_xfer++]);
+
+			/* After sending all bytes for current target, move on to the next target */
+			if (target->num_xfer == target->data_len) {
```

Comment:
num_xfer can sometimes not equal the `data_len` because in I3C a target can send a EoD before the expected `data_len` such as with GETCAPS and GETMXDS which can have a variable length. Does this handle this situation?